sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/meta/provider/SchemaCapableIOTableProviderWrapper.java

```diff
@@ -0,0 +1,132 @@
+import org.apache.beam.sdk.extensions.sql.meta.BaseBeamTable;
+import org.apache.beam.sdk.extensions.sql.meta.BeamSqlTable;
+import org.apache.beam.sdk.extensions.sql.meta.Table;
+import org.apache.beam.sdk.options.PipelineOptions;
+import org.apache.beam.sdk.schemas.Schema;
+import org.apache.beam.sdk.schemas.io.InvalidConfigurationException;
+import org.apache.beam.sdk.schemas.io.InvalidSchemaException;
+import org.apache.beam.sdk.schemas.io.SchemaIO;
+import org.apache.beam.sdk.schemas.io.SchemaIOProvider;
+import org.apache.beam.sdk.transforms.PTransform;
+import org.apache.beam.sdk.util.RowJson;
+import org.apache.beam.sdk.values.PBegin;
+import org.apache.beam.sdk.values.PCollection;
+import org.apache.beam.sdk.values.POutput;
+import org.apache.beam.sdk.values.Row;
+
+/**
+ * A general {@link TableProvider} for IOs for consumption by Beam SQL.
+ *
+ * <p>Can create child classes for IOs to pass {@link #getSchemaIOProvider()} that is specific to
+ * the IO.
+ */
+@Internal
+@Experimental
+public abstract class SchemaCapableIOTableProviderWrapper extends InMemoryMetaTableProvider
+    implements Serializable {
+  public abstract SchemaIOProvider getSchemaIOProvider();
+
+  @Override
+  public String getTableType() {
+    return getSchemaIOProvider().identifier();
+  }
+
+  @Override
+  public BeamSqlTable buildBeamSqlTable(Table tableDefinition) {
+    JSONObject tableProperties = tableDefinition.getProperties();
+
+    try {
+      RowJson.RowJsonDeserializer deserializer =
+          RowJson.RowJsonDeserializer.forSchema(getSchemaIOProvider().configurationSchema())
+              .withNullBehavior(RowJson.RowJsonDeserializer.NullBehavior.ACCEPT_MISSING_OR_NULL);
+
+      Row configurationRow =
+          newObjectMapperWith(deserializer).readValue(tableProperties.toString(), Row.class);
+
+      SchemaIO schemaIO =
+          getSchemaIOProvider()
+              .from(tableDefinition.getLocation(), configurationRow, tableDefinition.getSchema());
+
+      return new SchemaIOTableWrapper(schemaIO);
+    } catch (InvalidConfigurationException | InvalidSchemaException e) {
+      throw new InvalidTableException(e.getMessage());
+    } catch (JsonProcessingException e) {
+      throw new AssertionError(
+          "Failed to re-parse TBLPROPERTIES JSON " + tableProperties.toString());
+    }
+  }
+
+  private BeamTableStatistics getTableStatistics(PipelineOptions options) {
+    if (isBounded().equals(PCollection.IsBounded.BOUNDED)) {
+      return BeamTableStatistics.BOUNDED_UNKNOWN;
+    }
+    return BeamTableStatistics.UNBOUNDED_UNKNOWN;
+  }
+
+  private PCollection.IsBounded isBounded() {
+    return getSchemaIOProvider().isBounded();
+  }
+
+  /** A generalized {@link Table} for IOs to create IO readers and writers. */
+  private class SchemaIOTableWrapper extends BaseBeamTable {
+    protected final SchemaIO schemaIO;
+
+    private SchemaIOTableWrapper(SchemaIO schemaIO) {
```

Comment:
The reason why the two helper functions are there is because some IO Table Providers have logic that is more complex than the general form (ie [TextTable](https://github.com/apache/beam/blob/9b0941945545e71a949649309e05e405ca73aea2/sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/meta/provider/text/TextTable.java#L73-L96)), and we were planning on overriding those methods within the TableProvider of that IO.